### ISSUE TITLE
Chain Function checking

### ISSUE THREAD

**Issue by sirinath** [typetools/checker-framework#4231]
I am looking for a chain function enforcement where return value of one function is passed to another function.

E.g.

1.

```java
class C {
   @ChainTo("C::f(int)")
   int g(int i) {} // result of g must be passed to f
   int f(@ChainFrom("C::g(int)") int i) {} // i can only have the result of g(int)
}

...

int x = g(10);
f(x);
```

2.

```java
class C extends @When("closed == false") WhenOpen { // WhenOpen can only be accessed when closed == false
   boolean closed = false;
   @Once // can only be called once
   @CallBefore("C::g()") // can be only called before another call
   @CallAfterConstructor // should be invoked after construction but before any other method
   void f() {}
   @CallUntil("C::h()") // can be only called until another call after which it cannot be called again
   void g() {}
   @When("closed == false")
   void h() {}
   @State(param="i == 5", state="closed == false", transition="closed=true") // Enforce what state can be touched and what state traditions are allowed in functions for given parameters
   void x(int i) {}
}
```

**Comment by mernst** [comment]
This is easy to do, and is described in the manual.  You can define a new type annotation, say `@GResult`:

```java
  @GResult int g(int i) {}
  int f(@GResult int i) {}
```

You probably want to make `@GResult int` unrelated to the default annotation on `int`.

You can [create your own type system](https://checkerframework.org/manual/#creating-a-checker), or you can use the [Fake Enum Checker](https://checkerframework.org/manual/#fenum-checker) or the [Subtyping Checker](https://checkerframework.org/manual/#subtyping-checker) without the need to write code.
